src/main/java/io/spokestack/spokestack/nlu/tensorflow/EncodedTokens.java

```diff
@@ -0,0 +1,101 @@
+package io.spokestack.spokestack.nlu.tensorflow;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A simple data class that represents both the text of tokens produced from a
+ * full string and identifiers associated with those tokens.
+ */
+public class EncodedTokens {
+
+    private final List<String> originalTokens;
+    private final List<Integer> ids;
+    private List<Integer> originalIndices;
+
+    /**
+     * Create a new instance.
+     *
+     * @param spaceSeparated The original string split on whitespace.
+     */
+    public EncodedTokens(String[] spaceSeparated) {
+        this.ids = new ArrayList<>();
+        this.originalTokens = Arrays.asList(spaceSeparated);
+    }
+
+    /**
+     * @return The identifiers associated with the tokens.
+     */
+    public List<Integer> getIds() {
+        return ids;
+    }
+
+    /**
+     * Store the array mapping indices in the token ID list to those tokens'
+     * indices in the original token array.
+     *
+     * @param indices The mapping array.
+     * @throws IllegalStateException if the supplied array does not match the
+     *                               length of the token ID list.
+     */
+    public void setOriginalIndices(List<Integer> indices) {
+        if (indices.size() != ids.size()) {
+            throw new IllegalStateException(
+                  "original indices and the token ID list must be the "
+                        + "same length!");
+        }
+        this.originalIndices = indices;
+    }
+
+    /**
+     * Add a series of token ids.
+     *
+     * @param tokenIds The token identifiers.
+     */
+    public void addTokenIds(List<Integer> tokenIds) {
+        this.ids.addAll(tokenIds);
+    }
+
+    /**
+     * Use the index map created during encoding to recover a section of the
+     * original string. Like {@link List#subList(int, int)}, this method uses a
+     * half-open interval: the range returned includes the token referenced by
+     * {@code start} but excludes the one referenced by {@code stop} (unless a
+     * previous subtoken was also mapped to that token; see below).
+     *
+     * <p>
+     * For example, if the string {@code "The quick brown fox"} resulted in the
+     * token IDs {@code [1, 2, 3, 4]}, a call to {@code decodeRange(0, 2)} would
+     * return the text "The quick".
+     * </p>
+     *
+     * <p>
+     * If the end of the interval falls in the middle of a token that has been
```

Comment:
It guards against a coding mistake of mistakenly specifying a too-narrow interval, but more importantly, given the way that the original tokens are tracked in `EncodedTokens`, it's impossible to not alter the interval in some way—you either have to omit the token you're in the middle of or return the whole thing. I figured the latter was preferable; this of course shouldn't actually happen in practice, especially since `decodeRange` isn't something the client app is likely to use.